A single framework for gen-kill and generic dataflow problems

ecstatic-morse opened the pull request:
This is the prototype implementation discussed in rust-lang/compiler-team#202. You can read a high-level description of it in [the proposal](https://hackmd.io/@39Qr_z9cQhasi25sGjmFnA/Skvd9rztS) for that design meeting. This would eventually supersede the existing `BitDenotation` interface.

r? @ghost 

cc @rust-lang/compiler (esp. @eddyb and @pnkfelix)

bors commented:
:umbrella: The latest upstream changes (presumably #65869) made this pull request unmergeable. Please resolve the merge conflicts.

JohnCSimon commented:
Ping from triage:
@ecstatic-morse Hi! is this PR still in progress?
thanks.

JohnCSimon commented:
Pinging from triage:
@ecstatic-morse Can you please provide status on this PR?
thanks.

ecstatic-morse commented:
This was discussed in the design meeting on Nov. 8 (rust-lang/compiler-team#202). The consensus was to move forward on a design similar to the prototype. I'll be actively working to merge this, hopefully with some input from the compiler team.

ecstatic-morse commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 75775ba92a97fca8a798ef4407f72b2be31f7f59 with merge 1e0b457f588f21c0dfef20d076a5bc1371571aa1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"75775ba92a97fca8a798ef4407f72b2be31f7f59","merge_sha":"1e0b457f588f21c0dfef20d076a5bc1371571aa1"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13518)
Build commit: 1e0b457f588f21c0dfef20d076a5bc1371571aa1 (`1e0b457f588f21c0dfef20d076a5bc1371571aa1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13518"},"merge_sha":"1e0b457f588f21c0dfef20d076a5bc1371571aa1"} -->

rust-timer commented:
Queued 1e0b457f588f21c0dfef20d076a5bc1371571aa1 with parent 3f07f1cd78432c9901788de2feafc0c2331edee3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3f07f1cd78432c9901788de2feafc0c2331edee3&end=1e0b457f588f21c0dfef20d076a5bc1371571aa1).

rust-timer commented:
Finished benchmarking try commit 1e0b457f588f21c0dfef20d076a5bc1371571aa1, [comparison URL](https://perf.rust-lang.org/compare.html?start=3f07f1cd78432c9901788de2feafc0c2331edee3&end=1e0b457f588f21c0dfef20d076a5bc1371571aa1).

ecstatic-morse commented:
Okay, this converts a commonly used dataflow analysis, `MaybeInitializedLocals` to the new framework to get an idea of usability/perf changes. The `Engine` implementation is now much simpler than the original since the choice of whether to cache block transfer functions is dynamic. This means we will be able to skip this for acyclic MIR in the future.

I've triggered a perf run and the results look okay but there is a slight regression in `wg-grammar`. I will try lifting the check for a cached block transfer function out of the loop to see if that helps. However, I think this is ready for an initial round of review.

r? @pnkfelix cc @rust-lang/compiler (feel free to reassign)

Here are the unresolved "details" questions from the design meeting that are pertinent to this PR along with my thoughts on each. We should resolve these questions before merging.
 - How do we feel about the `{before_,}*_effect` naming convention? It takes a bit of getting used to, and doesn’t extend nicely to backwards dataflow.
    - OTOH, having a canonical `statement_effect` which most implementers want to use and a secondary, prefixed effect which is only rarely used is good.
- Should we be passing a `mir::Statement` to the `statement` effect methods? What about `mir::Body` to `bits_per_block`?
    - I'd like to have a `HasBody<'mir, 'tcx>` trait that is a requirement for `AnalysisDomain` instead of passing in a separate body explicitly, but this will increase the amount of boilerplate required.
- Should the `Analysis` methods take `&self` or `&mut self`?
    - Some may depend on the results of an earlier dataflow analysis and want to hold a cursor.
    - However, transfer functions should be pure for a given MIR, so maybe forcing those analyses to use interior mutability is appropriate.

ecstatic-morse commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit dceb65434dffaa017990114f9659feb8dcfedea7 with merge 43d649e19deb1597a734b28389fc636e8f1f2ad0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dceb65434dffaa017990114f9659feb8dcfedea7","merge_sha":"43d649e19deb1597a734b28389fc636e8f1f2ad0"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13596)
Build commit: 43d649e19deb1597a734b28389fc636e8f1f2ad0 (`43d649e19deb1597a734b28389fc636e8f1f2ad0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13596"},"merge_sha":"43d649e19deb1597a734b28389fc636e8f1f2ad0"} -->

rust-timer commented:
Queued 43d649e19deb1597a734b28389fc636e8f1f2ad0 with parent cefcc20898959a0c4fc11637cf8e53d3b6291125, future [comparison URL](https://perf.rust-lang.org/compare.html?start=cefcc20898959a0c4fc11637cf8e53d3b6291125&end=43d649e19deb1597a734b28389fc636e8f1f2ad0).

rust-timer commented:
Finished benchmarking try commit 43d649e19deb1597a734b28389fc636e8f1f2ad0, [comparison URL](https://perf.rust-lang.org/compare.html?start=cefcc20898959a0c4fc11637cf8e53d3b6291125&end=43d649e19deb1597a734b28389fc636e8f1f2ad0).

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13644&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13644/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13644/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-15T18:33:21.7769582Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-15T18:33:21.7957723Z ##[command]git config gc.auto 0
2019-11-15T18:33:21.8029097Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-15T18:33:21.8082668Z ##[command]git config --get-all http.proxy
2019-11-15T18:33:21.8231038Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65672/merge:refs/remotes/pull/65672/merge
---
2019-11-15T18:44:29.6775412Z    Compiling syntax_expand v0.0.0 (/checkout/src/libsyntax_expand)
2019-11-15T18:45:35.9147402Z    Compiling rustc_typeck v0.0.0 (/checkout/src/librustc_typeck)
2019-11-15T18:49:07.7662333Z    Compiling rustc_lint v0.0.0 (/checkout/src/librustc_lint)
2019-11-15T18:49:44.5945583Z    Compiling rustc_mir v0.0.0 (/checkout/src/librustc_mir)
2019-11-15T18:49:45.8129452Z error[E0425]: cannot find value `dead_unwinds` in this scope
2019-11-15T18:49:45.8129820Z    --> src/librustc_mir/dataflow/generic/engine.rs:138:18
2019-11-15T18:49:45.8130421Z     |
2019-11-15T18:49:45.8130663Z 138 |                 &dead_unwinds,
2019-11-15T18:49:45.8131005Z     |                  ^^^^^^^^^^^^ help: you might have meant to use the available field: `self.dead_unwinds`
2019-11-15T18:49:45.8177868Z error[E0425]: cannot find value `dead_unwinds` in this scope
2019-11-15T18:49:45.8177868Z error[E0425]: cannot find value `dead_unwinds` in this scope
2019-11-15T18:49:45.8178249Z    --> src/librustc_mir/dataflow/generic/engine.rs:146:18
2019-11-15T18:49:45.8178502Z     |
2019-11-15T18:49:45.8178746Z 146 |                 &dead_unwinds,
2019-11-15T18:49:45.8179084Z     |                  ^^^^^^^^^^^^ help: you might have meant to use the available field: `self.dead_unwinds`
2019-11-15T18:49:45.8179152Z 
2019-11-15T18:49:46.5974069Z error[E0687]: lifetimes used in `fn` or `Fn` syntax must be explicitly declared using `<...>` binders
2019-11-15T18:49:46.5974446Z    --> src/librustc_mir/dataflow/generic/engine.rs:194:87
2019-11-15T18:49:46.5974777Z     |
2019-11-15T18:49:46.5975302Z 194 |     apply_block_effect: impl Fn(&mut BitSet<A::Idx>, BasicBlock, &mir::BasicBlockData<'tcx>),
2019-11-15T18:49:46.5975728Z     |                                                                                       ^^^^ in-band lifetime definition
2019-11-15T18:49:46.7384273Z error: aborting due to 3 previous errors
2019-11-15T18:49:46.7384367Z 
2019-11-15T18:49:46.7389179Z For more information about this error, try `rustc --explain E0425`.
2019-11-15T18:49:46.7967745Z error: could not compile `rustc_mir`.
---
2019-11-15T18:51:38.1849393Z   local time: Fri Nov 15 18:51:38 UTC 2019
2019-11-15T18:51:38.4638204Z   network time: Fri, 15 Nov 2019 18:51:38 GMT
2019-11-15T18:51:38.4639013Z == end clock drift check ==
2019-11-15T18:51:41.2185168Z 
2019-11-15T18:51:41.2285902Z ##[error]Bash exited with code '1'.
2019-11-15T18:51:41.2318316Z ##[section]Starting: Checkout
2019-11-15T18:51:41.2319915Z ==============================================================================
2019-11-15T18:51:41.2319980Z Task         : Get sources
2019-11-15T18:51:41.2320036Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14705&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14705/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14705/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-28T23:10:36.8905791Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-28T23:10:37.6676485Z ##[command]git config gc.auto 0
2019-11-28T23:10:37.6682748Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-28T23:10:37.6687786Z ##[command]git config --get-all http.proxy
2019-11-28T23:10:37.6693394Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65672/merge:refs/remotes/pull/65672/merge
---
2019-11-28T23:19:55.6071588Z    Compiling rustc_macros v0.1.0 (/checkout/src/librustc_macros)
2019-11-28T23:20:06.6668633Z    Compiling syntax_pos v0.0.0 (/checkout/src/libsyntax_pos)
2019-11-28T23:20:08.3983350Z    Compiling rustc_errors v0.0.0 (/checkout/src/librustc_errors)
2019-11-28T23:20:28.9205156Z    Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
2019-11-28T23:20:58.8894123Z error[E0560]: struct `mir::Body<'_>` has no field named `__upvar_debuginfo_codegen_only_do_not_use`
2019-11-28T23:20:58.8895785Z     |
2019-11-28T23:20:58.8895785Z     |
2019-11-28T23:20:58.8896360Z 218 |             __upvar_debuginfo_codegen_only_do_not_use: Vec::new(),
2019-11-28T23:20:58.8896952Z     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `mir::Body<'_>` does not have this field
2019-11-28T23:20:58.8897409Z     |
2019-11-28T23:20:58.8897952Z     = note: available fields are: `basic_blocks`, `phase`, `source_scopes`, `source_scope_local_data`, `yield_ty` ... and 10 others
2019-11-28T23:21:08.1098416Z error: aborting due to previous error
2019-11-28T23:21:08.1099367Z 
2019-11-28T23:21:08.1104796Z For more information about this error, try `rustc --explain E0560`.
2019-11-28T23:21:08.2784586Z error: could not compile `rustc`.
---
2019-11-28T23:21:10.4011575Z   local time: Thu Nov 28 23:21:10 UTC 2019
2019-11-28T23:21:10.5491418Z   network time: Thu, 28 Nov 2019 23:21:10 GMT
2019-11-28T23:21:10.5495105Z == end clock drift check ==
2019-11-28T23:21:11.6216517Z 
2019-11-28T23:21:11.6329596Z ##[error]Bash exited with code '1'.
2019-11-28T23:21:11.6366115Z ##[section]Starting: Checkout
2019-11-28T23:21:11.6368656Z ==============================================================================
2019-11-28T23:21:11.6368725Z Task         : Get sources
2019-11-28T23:21:11.6368778Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14932&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14932/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14932/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-01T20:33:15.3358547Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-01T20:33:15.3547362Z ##[command]git config gc.auto 0
2019-12-01T20:33:16.3159980Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-01T20:33:16.3161821Z ##[command]git config --get-all http.proxy
2019-12-01T20:33:16.3163953Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65672/merge:refs/remotes/pull/65672/merge
---
2019-12-01T20:41:15.6117281Z    Compiling syntax_pos v0.0.0 (/checkout/src/libsyntax_pos)
2019-12-01T20:41:16.9836583Z    Compiling rustc_feature v0.0.0 (/checkout/src/librustc_feature)
2019-12-01T20:41:18.3510535Z    Compiling rustc_errors v0.0.0 (/checkout/src/librustc_errors)
2019-12-01T20:41:35.2673804Z    Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
2019-12-01T20:41:58.0327690Z error[E0063]: missing fields `generator_kind`, `var_debug_info` in initializer of `mir::Body<'_>`
2019-12-01T20:41:58.0329067Z     |
2019-12-01T20:41:58.0329684Z 213 |         Body {
2019-12-01T20:41:58.0329684Z 213 |         Body {
2019-12-01T20:41:58.0330195Z     |         ^^^^ missing `generator_kind`, `var_debug_info`
2019-12-01T20:42:05.5345893Z error: aborting due to previous error
2019-12-01T20:42:05.5348796Z 
2019-12-01T20:42:05.5356348Z For more information about this error, try `rustc --explain E0063`.
2019-12-01T20:42:05.6793672Z error: could not compile `rustc`.
---
2019-12-01T20:42:07.2644274Z   local time: Sun Dec  1 20:42:07 UTC 2019
2019-12-01T20:42:07.5400495Z   network time: Sun, 01 Dec 2019 20:42:07 GMT
2019-12-01T20:42:07.5403717Z == end clock drift check ==
2019-12-01T20:42:08.5438702Z 
2019-12-01T20:42:08.5504676Z ##[error]Bash exited with code '1'.
2019-12-01T20:42:08.5527912Z ##[section]Starting: Checkout
2019-12-01T20:42:08.5529291Z ==============================================================================
2019-12-01T20:42:08.5529346Z Task         : Get sources
2019-12-01T20:42:08.5529381Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #64736) made this pull request unmergeable. Please resolve the merge conflicts.

JohnCSimon commented:
Ping from triage: 
@ecstatic-morse - can you please fix the merge conflicts?

ecstatic-morse commented:
@pnkfelix This is waiting for review. Feel free to reassign it. I could split out the unit test for `ResultsCursor` and the conversion of `MaybeInitializedPlaces` to the new framework if the diff is too imposing (I included it here so I could measure perf).

ecstatic-morse commented:
@bors try
@rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit a9d46c7dafbe82f4dc2d341c624cd1eb63309c80 with merge d05d26499af24c8153dd8f16bb860abe185d18b2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a9d46c7dafbe82f4dc2d341c624cd1eb63309c80","merge_sha":"d05d26499af24c8153dd8f16bb860abe185d18b2"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=15576)
Build commit: d05d26499af24c8153dd8f16bb860abe185d18b2 (`d05d26499af24c8153dd8f16bb860abe185d18b2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=15576"},"merge_sha":"d05d26499af24c8153dd8f16bb860abe185d18b2"} -->

rust-timer commented:
Queued d05d26499af24c8153dd8f16bb860abe185d18b2 with parent 76a252ea9e7be93a61ffdf33b3533e24a9cf459d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=76a252ea9e7be93a61ffdf33b3533e24a9cf459d&end=d05d26499af24c8153dd8f16bb860abe185d18b2).

rust-timer commented:
Finished benchmarking try commit d05d26499af24c8153dd8f16bb860abe185d18b2, [comparison URL](https://perf.rust-lang.org/compare.html?start=76a252ea9e7be93a61ffdf33b3533e24a9cf459d&end=d05d26499af24c8153dd8f16bb860abe185d18b2).

ecstatic-morse commented:
Seems there's no perf info for this this commit.

@rust-timer queue

rust-timer commented:
Awaiting bors try build completion

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit a9d46c7dafbe82f4dc2d341c624cd1eb63309c80 with merge 5c014acfbbdd9154c551cd697f21f1132fc7aa8e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a9d46c7dafbe82f4dc2d341c624cd1eb63309c80","merge_sha":"5c014acfbbdd9154c551cd697f21f1132fc7aa8e"} -->

lqd commented:
@ecstatic-morse they are in the queue — https://perf.rust-lang.org/status.html — the server had an issue earlier and some data needed to be recomputed IIUC.

ecstatic-morse commented:
Whoops, I see it now.

@bors try-

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

pnkfelix commented on src/librustc_mir/dataflow/generic/mod.rs:
```diff
@@ -0,0 +1,305 @@
+
+/// The legal operations for a transfer function in a gen/kill problem.
+pub trait GenKill<T>: Sized {
+    /// Inserts `elem` into the `gen` set, removing it the `kill` set if present.
```
typo: removing it *from* the `kill` set if present

pnkfelix replied:
also, the phrase "if present" is slightly ambiguous. 

When I first read "if present", I thought the phrase ("clearly") was referring to the `elem`, or rather the bit representing `elem` in the `kill` set. Which made sense to me.

But now that I have read the whole commit, I realize that it is possible that the "if present" was actually referring o the whole `kill`-set: i.e., this was an implicit note that the `kill`-set might be missing. (And likewise, the "if present" in `fn kill` below might be referring to the whole `gen`-set.)

I do not know what your actual intention was. I have already stated that I think it is a mistake to have `impl GenKill for BitSet`. If your intention matched my original interpretation, then maybe you can rephrase this as "Inserts `elem` into the `gen` set, removing it from the `kill` set if `elem` was present there." Or you could just say "Inserts `elem` into the `gen` set and removes it from the `kill` set", and rely on the reader to understand that under usual set theory, removing a non-present element is the identity function on sets.

pnkfelix commented on src/librustc_mir/dataflow/generic/mod.rs:
```diff
@@ -0,0 +1,305 @@
+    /// Inserts `elem` into the `gen` set, removing it the `kill` set if present.
+    fn gen(&mut self, elem: T);
+
+    /// Inserts `elem` into the `kill` set, removing it the `gen` set if present.
```
same typo: removing it *from* the `gen` set if present

pnkfelix commented on src/librustc_mir/dataflow/generic/mod.rs:
```diff
@@ -0,0 +1,305 @@
+}
+
+/// The legal operations for a transfer function in a gen/kill problem.
+pub trait GenKill<T>: Sized {
```
I'm only on the first commit, so I don't really know the answer to this question yet: Does abstracting `GenKill<T>` into a trait make sense? Its hard to imagine a significantly different alternative implementation to what the existing `GenKillSet` struct does, apart from the details of how the two bitsets are represented, I suppose.

pnkfelix replied:
Okay I reached the end of this commit and now see that you have implemented `GenKill` for `BitSet`. I do not think this is a good idea.

Specifically, you have not stated in the documentation for `trait GenKill` that the underlying gen or kill sets may be *missing*, depending on how the trait is implemented. Having `fn kill()` just remove the bit from the gen-set does not really capture the semantics of a dataflow analysis: the clients parameterized over `GK: GenKill` are going to have to say "You must use `GenKillSet` for this to work", or they are going to have to ensure that their particular analysis does not actually rely ever generating a kill effect (in which case, I would have `<BitSet as GenKill>::kill` panic instead of solely clearing the bit in the gen-set.

pnkfelix replied:
(I suppose another way it could hypothetically make sense to `impl GenKill for BitSet` would be for an analysis where the kill-set is *always* the set-complement of the gen-set. But AFAICT that is not what you implemented, and I am not aware of any analysis where the gen/kill sets satisfy that property.)

ecstatic-morse replied:
In `ResultsCursor`, the `apply` methods are applied directly to the `BitSet` containing the dataflow state at the cursor position. The `GenKill` trait is meant to abstract over two different contexts where we need to apply the transfer function. When coalescing transfer functions for an entire block before running dataflow analysis, we are storing a function (`trans`) that can be applied later, which requires two bitsets. However, when seeking with a dataflow cursor, we don't care about the function itself; we only need to know its effect on the state vector, (`trans(state)`). There's no need to store extra bitsets.

There are other ways to avoid allocating here, but having a proper trait is the cleanest approach IMO. I agree that the documentation should be clearer about what is happening when in the `GenKill` implementation for `BitSet`.

pnkfelix replied:
Okay. I'm sorry that I did not read this response more carefully back when you first posted it, as it would have saved me the effort of rediscovering the motivation for this abstraction myself. (But then again, it did force you to embed the motivation into the doc-comment for the trait, so maybe my oversight was a win in the long run...)

pnkfelix commented on src/librustc_mir/dataflow/move_paths/mod.rs:
```diff
@@ -72,6 +72,40 @@ impl<'tcx> MovePath<'tcx> {
         parents
     }
+
+    /// Looks for the first child of `self` for which `f` returns `true`.
```
I often read the term "child" as meaning "immediate child", as opposed to recursively traversing the family tree.

 1. Would "descendant" be better?
 2. Regardless of whether you change the terminology, you should state in the documentation the search strategy. I.e., If you say "Looks for the first descendant (under breadth-first search) for which ...", then I think it will be 100% clear that this is not restricting its search to the immediate children (because one would not distinguish DFS/BFS for a traversal of the immediate children).

pnkfelix commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -185,15 +185,11 @@ fn do_mir_borrowck<'a, 'tcx>(
+    let mut flow_inits =
+        dataflow::generic::Engine::new_gen_kill(tcx, &body, def_id, &dead_unwinds, flow_inits)
+            .iterate_to_fixpoint()
+            .into_cursor(&body);
```
OMG this is so much nicer to read than what I had before

pnkfelix commented on src/librustc_mir/dataflow/generic/tests.rs:
```diff
@@ -0,0 +1,297 @@
+        }
+    }
+
+    /// Returns the largest index (not including the basic block index) that should be set when
```
I'm trying to understand the docs here, and this method + test-harness in general.

 * I have inferred that the indexes you are talking about are points before and after statements: { } S1 { }, { } S2 { }, ... etc. (At first I wasn't sure why you were distinguishing the point after S1 from the point before S2, but soon I realized that you are identifying points in the whole MIR, not just within a given basic block, and therefore in general you have to deal with control-flow splits and merges.)

 * When you say "that should be set", where are these indices being "*set*"? Is that referring to the `Dummy` that is storing the indices in a set?

 * Why is this called `effect` ? Because it is describing the "effect" of calling `seek` with a given `SeekTarget`? Or just that it describes the effect we want to compute on the bitsets for `Dummy`?

(I suppose part of my problem here is that I started reading this file from top to bottom, and it seems like the important explanations are further down, with `Dummy` and its related impls.)

ecstatic-morse replied:
I'll start from first principles since I didn't do a good job of explaining the motivation for this.

The ultimate goal of this is to test the logic that updates the internal state of a `dataflow::ResultsCursor` (this is where indices should be "set") by seeking *to* every possible target (before or after a `Location`) *from* every possible target. For this test to be exhaustive, we must define a dataflow analysis such that every possible effect (before, after and call-return) is unique for every statement and terminator in a MIR body. This ensures that the expected state at every possible target is unique over the entire MIR body. There's no need to actually find the fixpoint for such an analysis to test `ResultsCursor`, so the actual CFG is inconsequential.

The actual implementation does not quite live up to this ideal. Instead it takes advantage of the fact that we never iterate to fixpoint by storing the index of each basic block in its entry set. The effect of each statement in our dummy analysis is only "unique" when considered in conjunction with the index that appears in the entry set.

I could reimplement this test it to be more like the first paragraph. Since I need to map before/after/call-return effects for each statement and terminator to a unique index, I'm afraid that it will always be a bit opaque. I could also keep it roughly as-is with an edited version of this comment as documentation, or remove it entirely.

pnkfelix commented on src/librustc_mir/dataflow/generic/tests.rs:
```diff
@@ -0,0 +1,297 @@
+    /// Returns the expected state at the given `SeekTarget`.
+    ///
+    /// This is the union of index of the target basic block, the index assigned to the of the
+    /// target statement or terminator, and the indices of all preceding statements in the MIR.
```
when you say "indices of all preceding statements in the MIR", does that mean:

 * all preceding statements *within* the target basic basic block (but not other blocks), or
 * all preceding statements both in the target basic block and elsewhere
    * if this, then is the "preceding" just talking about the order the statements are given in the MIR array,
    * or does "preceding" refer to control flow in some manner (but presumably some manner that ignores backward edges)?

ecstatic-morse replied:
I meant the first interpretation. This was unclear.

pnkfelix commented on src/librustc_mir/dataflow/generic/tests.rs:
```diff
@@ -0,0 +1,297 @@
+        projection: ty::List::empty(),
+    };
+
+    block(4, mir::TerminatorKind::Return);
```
I'm surprised there's no block with 0 statements (i.e. just a terminator) in included in this test.

pnkfelix commented on src/librustc_mir/dataflow/generic/tests.rs:
```diff
@@ -0,0 +1,297 @@
+    /// target statement or terminator, and the indices of all preceding statements in the MIR.
+    ///
+    /// For example, the expected state when calling
+    /// `seek_before(Location { block: 2, statement_index: 2 })` would be `[102, 0, 1, 2, 3, 4]`.
```
I spent a little while reading over the dummy_body()  cfg construction code and tried to understand why that control-flow graph would yield the results `[102, 0, 1, 2, 3, 4]`, but at some point I decided that if I have to work this hard to figure it out, then it needs to be explained more explicitly in the documentation here. In short, I don't know why I should expect the aforementioned output for that case. (e.g. Where does the 102 come from?)

pnkfelix commented on src/librustc_mir/dataflow/generic/engine.rs:
```diff
@@ -167,124 +156,155 @@ where
-    fn apply_whole_block_effect(
+    /// Helper function that propagates dataflow state into graph succesors until fixpoint is
+    /// reached.
+    fn _iterate_to_fixpoint(
```
eek, I don't think you should name helpers with a *leading* underscore. Leading underscore is usually used for fields/methods that might be dead code.

I'd personally use a word to differentiate the helper routiine (e.g. `iterate_to_fixpoint_core`) but even just a *trailing* underscore would be better here than a leading underscore.

ecstatic-morse replied:
This is my personal coding convention that probably originates from Python. I will switch this (as well as `_seek`) to a trailing underscore unless there is an established convention.

pnkfelix commented:
Okay. Overall this looks good. My main complaint is about the `impl GenKill for BitSet`. Are you using that impl? If not, then I'd say get rid of it (and potentially get rid of `trait GenKill` entirely, at least for now.

pnkfelix commented:
Regarding the specific questions raised by @ecstatic-morse : 

> - How do we feel about the `{before_,}*_effect` naming convention? It takes a bit of getting used to, and doesn’t extend nicely to backwards dataflow.
>    - OTOH, having a canonical `statement_effect` which most implementers want to use and a secondary, prefixed effect which is only rarely used is good.

While reading the commit series, I definitely asked myself "can we get rid of the `before_{statement, terminator}_effect` methods entirely? Its not clear what purpose they currently serve, because they are always called (I think) immediately before their corresponding partner `{statement, terminator}_effect` method, right? (Or is there some context where only the latter is called without the former? I guess I would need to review the code more thoroughly.

In any case, I would be fine with leaving the convention as is for now, and making a follow-up task to revisit this part of the API. I'd be thrilled to clean it up, but I wouldn't block landing this PR on that task.

> - Should we be passing a `mir::Statement` to the `statement` effect methods? What about `mir::Body` to `bits_per_block`?
>    - I'd like to have a `HasBody<'mir, 'tcx>` trait that is a requirement for `AnalysisDomain` instead of passing in a separate body around everywhere, but this will increase the amount of boilerplate required.

I'm fine with passing in the `Statement`; it seems like most effects will need that info, so why make all the clients do the lookup, right? I don't care as much either way about whether or not we pass in the `Body` to `bits_per_block` (I don't know offhand, but I would assume a similar argument applies there that most consumers will need to extract info from the `Body` in order to determine the size of their domain's bitvector.)

> - Should the `Analysis` methods take `&self` or `&mut self`?
>     - Some may depend on the results of an earlier dataflow analysis and want to hold a cursor.
>     - However, transfer functions should be pure for a given MIR, so maybe forcing those analyses to use interior mutability is appropriate.

I'm torn. I would vote for `&self` here, personally. But I know first-hand the pain of trying to wedge in an analysis-dependent analysis after the framework was originally designed without that use case in mind, and your argument there for `&mut self` is at least giving me pause.

Is there some semi-standard way we could set things up so that callers of `Analysis` see its API as `&mut self`, while also having implementors of `Analysis` see its API as `&self`, so that we would force each side to be conservative in their implementation choices while delaying committing ourselves to one side or another right now?

ecstatic-morse commented:
> While reading the commit series, I definitely asked myself "can we get rid of the before_{statement, terminator}_effect methods entirely? Its not clear what purpose they currently serve, because they are always called (I think) immediately before their corresponding partner {statement, terminator}_effect method, right? (Or is there some context where only the latter is called without the former? I guess I would need to review the code more thoroughly.)

The only implementer of the "before" variants is in [`src/librustc_mir/dataflow/impls/borrows.rs`](https://github.com/rust-lang/rust/blob/25434f898b499876203a3b95c1b38bad5ed2cc5d/src/librustc_mir/dataflow/impls/borrows.rs#L249). They could almost be factored out by incorporating the "before" effect of statement `i` into the "after" effect of statement `i-1`. However, this won't work if the first statement in a block has a meaningful "before" effect. This may not be the case for the `Borrows` analysis, since what loans would go out of scope before the first statement?

ecstatic-morse commented:
Ideally, this PR would be performance neutral, but there's a slight regression on [two benchmarks](https://perf.rust-lang.org/compare.html?start=76a252ea9e7be93a61ffdf33b3533e24a9cf459d&end=d05d26499af24c8153dd8f16bb860abe185d18b2). I currently have a micro-optimization in this PR that partially addresses this. I would like to split that optimization out into a later PR that also ports existing analyses to the new framework (importantly `MaybeInitializedPlaces`, `MaybeUninitializedPlaces`, `Borrows`) to the new framework so I can get a fuller picture of its performance. This PR only includes a port of the first, which means the perf diff is more susceptible to noise.

ecstatic-morse commented:
> 

Is there some semi-standard way we could set things up so that callers of Analysis see its API as &mut self, while also having implementors of Analysis see its API as &self, so that we would force each side to be conservative in their implementation choices while delaying committing ourselves to one side or another right now?

I suppose we could have an `AnalysisMut` facade that is used by the `engine` and `cursor`. I think it's not worth the extra complexity. I'd rather just keep the status quo and fix any borrow check errors in `engine` and `cursor` by hand if we decide we want to switch to `&mut analysis`.

bors commented:
:umbrella: The latest upstream changes (presumably #67707) made this pull request unmergeable. Please resolve the merge conflicts.

ecstatic-morse commented:
@pnkfelix I reorganized the unit test and added docs to `GenKill` and `GenKillSet` to explain how they should be used. I've moved the migration of `MaybeInitializedPlaces` to the new framework into a separate branch so that I can work on optimizing them without pushing a bunch of commits to this PR.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=18461&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/18461/logs/70), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/18461/logs/70)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-01-15T00:14:52.4458606Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-01-15T00:14:52.4537165Z ##[command]git config gc.auto 0
2020-01-15T00:14:52.4609160Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-01-15T00:14:52.4662296Z ##[command]git config --get-all http.proxy
2020-01-15T00:14:52.4793782Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65672/merge:refs/remotes/pull/65672/merge
---
2020-01-15T01:06:44.0912170Z .................................................................................................... 1600/9519
2020-01-15T01:06:48.9780142Z .................................................................................................... 1700/9519
2020-01-15T01:06:57.1097242Z .................................................................................................... 1800/9519
2020-01-15T01:07:05.3047219Z ........i........................................................................................... 1900/9519
2020-01-15T01:07:11.6122117Z ..................................................................................................ii 2000/9519
2020-01-15T01:07:27.1239016Z iii................................................................................................. 2100/9519
2020-01-15T01:07:35.1434384Z .................................................................................................... 2300/9519
2020-01-15T01:07:37.4984402Z .................................................................................................... 2400/9519
2020-01-15T01:07:43.1200540Z .................................................................................................... 2500/9519
2020-01-15T01:08:02.0381055Z .................................................................................................... 2600/9519
---
2020-01-15T01:10:30.2007711Z .........................................i...............i.......................................... 4900/9519
2020-01-15T01:10:38.6669918Z .................................................................................................... 5000/9519
2020-01-15T01:10:44.6236966Z ....................................................................................i............... 5100/9519
2020-01-15T01:10:49.6892156Z .................................................................................................... 5200/9519
2020-01-15T01:10:59.3831938Z .......................................................ii.ii...........i............................ 5300/9519
2020-01-15T01:11:08.3027988Z .................................................................................................... 5500/9519
2020-01-15T01:11:17.9719515Z .................................................................................................... 5600/9519
2020-01-15T01:11:24.1187843Z ........................................i........................................................... 5700/9519
2020-01-15T01:11:30.5490924Z .................................................................................................... 5800/9519
2020-01-15T01:11:30.5490924Z .................................................................................................... 5800/9519
2020-01-15T01:11:40.6005580Z .................................................................................................... 5900/9519
2020-01-15T01:11:49.6945761Z ...............................ii...i..ii...........i............................................... 6000/9519
2020-01-15T01:12:07.9742991Z .................................................................................................... 6200/9519
2020-01-15T01:12:14.9211306Z .................................................................................................... 6300/9519
2020-01-15T01:12:14.9211306Z .................................................................................................... 6300/9519
2020-01-15T01:12:26.6724592Z ...........................................................i..ii.................................... 6400/9519
2020-01-15T01:12:53.3995890Z .................................................................................................... 6600/9519
2020-01-15T01:12:55.3106803Z ...................................i................................................................ 6700/9519
2020-01-15T01:12:57.9848795Z .................................................................................................... 6800/9519
2020-01-15T01:12:59.5913981Z ...................................i................................................................ 6900/9519
---
2020-01-15T01:14:29.4334295Z .................................................................................................... 7500/9519
2020-01-15T01:14:33.4081568Z .................................................................................................... 7600/9519
2020-01-15T01:14:38.8966165Z .................................................................................................... 7700/9519
2020-01-15T01:14:45.4689874Z .................................................................................................... 7800/9519
2020-01-15T01:14:54.5657115Z ....................................................................................iiii............ 7900/9519
2020-01-15T01:15:09.8445490Z ..................i......i.......................................................................... 8100/9519
2020-01-15T01:15:14.7285925Z .................................................................................................... 8200/9519
2020-01-15T01:15:27.2680196Z .................................................................................................... 8300/9519
2020-01-15T01:15:36.7526753Z .................................................................................................... 8400/9519
---
2020-01-15T01:17:52.1550503Z  finished in 6.506
2020-01-15T01:17:52.1733045Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:17:52.3634808Z 
2020-01-15T01:17:52.3635676Z running 166 tests
2020-01-15T01:17:55.1016026Z iiii......i........ii..iiii...i....i...........i............i..i..................i....i............ 100/166
2020-01-15T01:17:57.2667430Z i.i.i...iii..iiiiiii.......................iii............ii......
2020-01-15T01:17:57.2674297Z 
2020-01-15T01:17:57.2679036Z  finished in 5.094
2020-01-15T01:17:57.2863245Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:17:57.4468706Z 
---
2020-01-15T01:17:59.2431724Z  finished in 1.957
2020-01-15T01:17:59.2597546Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:17:59.4122810Z 
2020-01-15T01:17:59.4122952Z running 9 tests
2020-01-15T01:17:59.4123785Z iiiiiiiii
2020-01-15T01:17:59.4124084Z 
2020-01-15T01:17:59.4132037Z  finished in 0.152
2020-01-15T01:17:59.4298509Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:17:59.6010352Z 
---
2020-01-15T01:18:18.0813167Z  finished in 18.651
2020-01-15T01:18:18.1018163Z Check compiletest suite=debuginfo mode=debuginfo-gdb+lldb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:18:18.2752160Z 
2020-01-15T01:18:18.2752946Z running 116 tests
2020-01-15T01:18:41.1939366Z .iiiii..i.....i..i...i..i.i.i..i..i..ii....i.i....ii..........iiii..........i.....i..i.......ii.i.ii 100/116
2020-01-15T01:18:44.3706497Z .....iiii.....ii
2020-01-15T01:18:44.3707402Z 
2020-01-15T01:18:44.3711048Z  finished in 26.269
2020-01-15T01:18:44.3717727Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-01-15T01:18:44.3718047Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2020-01-15T01:30:46.8804896Z 
2020-01-15T01:30:46.8806198Z    Doc-tests core
2020-01-15T01:30:51.2012019Z 
2020-01-15T01:30:51.2013564Z running 2442 tests
2020-01-15T01:30:59.8264030Z ......iiiii......................................................................................... 100/2442
2020-01-15T01:31:08.5881431Z ..................................................................................ii................ 200/2442
2020-01-15T01:31:28.6603973Z ................i................................................................................... 400/2442
2020-01-15T01:31:28.6603973Z ................i................................................................................... 400/2442
2020-01-15T01:31:37.5409545Z .................................................................i..i..................iiii......... 500/2442
2020-01-15T01:31:52.9765005Z .................................................................................................... 700/2442
2020-01-15T01:32:01.1042984Z .................................................................................................... 800/2442
2020-01-15T01:32:09.1374827Z .................................................................................................... 900/2442
2020-01-15T01:32:17.2436204Z .................................................................................................... 1000/2442
---
2020-01-15T01:35:27.8719513Z .................................................................................................... 500/760
2020-01-15T01:35:27.9077019Z .................thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2741:22
2020-01-15T01:35:27.9090522Z ....thread 'thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2778:21
2020-01-15T01:35:27.9093603Z <unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', src/libstd/sync/mpsc/mod.rs:2766:17
2020-01-15T01:35:27.9111826Z .......thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2645:13
2020-01-15T01:35:28.1684542Z ..........................................thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:1997:22
2020-01-15T01:35:28.1711645Z ....thread '.<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs.:2034:21
2020-01-15T01:35:28.1732662Z .....thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:1916:13
2020-01-15T01:35:30.2141602Z ...................thread '<unnamed>' panicked at 'explicit panic', src/libstd/sync/mutex.rs:633:13
2020-01-15T01:35:30.2146869Z ..thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:587:13
2020-01-15T01:35:30.2153217Z thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:563:13
2020-01-15T01:35:30.2167151Z ....thread '<unnamed>' panicked at 'explicit panic', src/libstd/sync/mutex.rs:694:13
---
2020-01-15T01:35:39.2847672Z 
2020-01-15T01:35:39.2847932Z running 1003 tests
2020-01-15T01:35:56.4925238Z i................................................................................................... 100/1003
2020-01-15T01:36:06.0719349Z .................................................................................................... 200/1003
2020-01-15T01:36:13.1641914Z ..................iii......i......i...i......i...................................................... 300/1003
2020-01-15T01:36:18.0126719Z .................................................................................................... 400/1003
2020-01-15T01:36:24.6469118Z ..........................................i..i.....................................ii............... 500/1003
2020-01-15T01:36:36.7960682Z .................................................................................................... 700/1003
2020-01-15T01:36:36.7960682Z .................................................................................................... 700/1003
2020-01-15T01:36:42.8972344Z .............................iiii................................................................... 800/1003
2020-01-15T01:36:56.1907781Z .................................................................................................... 900/1003
2020-01-15T01:37:02.7650975Z ...................................................iiii............................................. 1000/1003
2020-01-15T01:37:02.8567393Z test result: ok. 983 passed; 0 failed; 20 ignored; 0 measured; 0 filtered out
2020-01-15T01:37:02.8567632Z 
2020-01-15T01:37:02.8675393Z  finished in 160.249
2020-01-15T01:37:02.8689834Z Testing term stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
2020-01-15T01:39:34.4113116Z 
2020-01-15T01:39:34.6578866Z error[E0425]: cannot find value `body` in this scope
2020-01-15T01:39:34.6579595Z    --> src/librustc_mir/dataflow/generic/tests.rs:101:61
2020-01-15T01:39:34.6580072Z     |
2020-01-15T01:39:34.6580582Z 101 |         let mut ret = BitSet::new_empty(self.bits_per_block(body));
2020-01-15T01:39:34.6581135Z     |                                                             ^^^^ help: you might have meant to use the available field: `self.body`
2020-01-15T01:39:34.6603473Z error[E0425]: cannot find value `body` in this scope
2020-01-15T01:39:34.6604655Z    --> src/librustc_mir/dataflow/generic/tests.rs:107:59
2020-01-15T01:39:34.6605250Z     |
2020-01-15T01:39:34.6605250Z     |
2020-01-15T01:39:34.6605897Z 107 |         let empty = BitSet::new_empty(self.bits_per_block(body));
2020-01-15T01:39:34.6606620Z     |                                                           ^^^^ help: you might have meant to use the available field: `self.body`
2020-01-15T01:39:34.6606904Z 
2020-01-15T01:39:38.8289263Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8290283Z    --> src/librustc_mir/dataflow/generic/tests.rs:186:47
2020-01-15T01:39:38.8290738Z     |
2020-01-15T01:39:38.8293345Z 186 |         let idx = SeekTarget::After(location).effect(self.body).unwrap();
2020-01-15T01:39:38.8293970Z     |                                               ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:38.8294581Z ...
2020-01-15T01:39:38.8295084Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8295475Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8295542Z 
2020-01-15T01:39:38.8363037Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8363301Z    --> src/librustc_mir/dataflow/generic/tests.rs:196:48
2020-01-15T01:39:38.8363503Z     |
2020-01-15T01:39:38.8363741Z 196 |         let idx = SeekTarget::Before(location).effect(self.body).unwrap();
2020-01-15T01:39:38.8365022Z     |                                                ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:38.8365294Z ...
2020-01-15T01:39:38.8365571Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8365910Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8365952Z 
2020-01-15T01:39:38.8435480Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8435854Z    --> src/librustc_mir/dataflow/generic/tests.rs:206:47
2020-01-15T01:39:38.8436112Z     |
2020-01-15T01:39:38.8436430Z 206 |         let idx = SeekTarget::After(location).effect(self.body).unwrap();
2020-01-15T01:39:38.8436852Z     |                                               ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:38.8437076Z ...
2020-01-15T01:39:38.8437335Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8437673Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8437715Z 
2020-01-15T01:39:38.8506075Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8506468Z    --> src/librustc_mir/dataflow/generic/tests.rs:216:48
2020-01-15T01:39:38.8506700Z     |
2020-01-15T01:39:38.8507020Z 216 |         let idx = SeekTarget::Before(location).effect(self.body).unwrap();
2020-01-15T01:39:38.8507446Z     |                                                ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:38.8507670Z ...
2020-01-15T01:39:38.8508079Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8508338Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8508369Z 
2020-01-15T01:39:38.8579354Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8580163Z    --> src/librustc_mir/dataflow/generic/tests.rs:229:64
2020-01-15T01:39:38.8580603Z     |
2020-01-15T01:39:38.8581214Z 229 |         let idx = SeekTarget::AfterAssumeCallReturns(location).effect(self.body).unwrap();
2020-01-15T01:39:38.8581848Z     |                                                                ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:38.8582257Z ...
2020-01-15T01:39:38.8582703Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8583175Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8583536Z 
2020-01-15T01:39:38.8695581Z error[E0599]: no method named `effect` found for enum `dataflow::generic::tests::SeekTarget` in the current scope
2020-01-15T01:39:38.8695964Z    --> src/librustc_mir/dataflow/generic/tests.rs:289:75
2020-01-15T01:39:38.8696211Z     |
2020-01-15T01:39:38.8696503Z 235 | enum SeekTarget {
2020-01-15T01:39:38.8696832Z     | --------------- method `effect` not found for this
2020-01-15T01:39:38.8697044Z ...
2020-01-15T01:39:38.8697584Z 289 |     assert_ne!(call_return_effect, SeekTarget::After(call_terminator_loc).effect(body).unwrap());
2020-01-15T01:39:38.8698078Z     |                                                                           ^^^^^^ method not found in `dataflow::generic::tests::SeekTarget`
2020-01-15T01:39:43.1982260Z error: aborting due to 9 previous errors
2020-01-15T01:39:43.1983121Z 
2020-01-15T01:39:43.1983638Z Some errors have detailed explanations: E0425, E0433, E0599.
2020-01-15T01:39:43.1984357Z For more information about an error, try `rustc --explain E0425`.
2020-01-15T01:39:43.1984357Z For more information about an error, try `rustc --explain E0425`.
2020-01-15T01:39:43.2168251Z error: could not compile `rustc_mir`.
2020-01-15T01:39:43.2169434Z 
2020-01-15T01:39:43.2169907Z To learn more, run the command again with --verbose.
2020-01-15T01:39:43.2182238Z 
2020-01-15T01:39:43.2182505Z 
2020-01-15T01:39:43.2183300Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-Zconfig-profile" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--features" " llvm" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "-p" "rustc_mir" "--" "--quiet"
2020-01-15T01:39:43.2183639Z 
2020-01-15T01:39:43.2183758Z 
2020-01-15T01:39:43.2191542Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2020-01-15T01:39:43.2191626Z Build completed unsuccessfully in 1:18:47
2020-01-15T01:39:43.2191626Z Build completed unsuccessfully in 1:18:47
2020-01-15T01:39:43.2243185Z == clock drift check ==
2020-01-15T01:39:43.2265138Z   local time: Wed Jan 15 01:39:43 UTC 2020
2020-01-15T01:39:43.5246185Z   network time: Wed, 15 Jan 2020 01:39:43 GMT
2020-01-15T01:39:43.5251781Z == end clock drift check ==
2020-01-15T01:39:44.0156124Z 
2020-01-15T01:39:44.0242567Z ##[error]Bash exited with code '1'.
2020-01-15T01:39:44.0280412Z ##[section]Starting: Checkout
2020-01-15T01:39:44.0282179Z ==============================================================================
2020-01-15T01:39:44.0282223Z Task         : Get sources
2020-01-15T01:39:44.0282435Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ecstatic-morse commented:
I migrated the analyses used during borrow checking/drop elaboration to the new framework and perf [seems good](https://github.com/rust-lang/rust/pull/68241#issuecomment-574570513). I'm gonna try a few small optimizations I mentioned during the design meeting over on that PR as well.

While doing the conversion, it became apparent that code would much be nicer if we extended the "builder" style all the way back to `Analysis` by adding an `into_engine` method to it. For example, code like this:

```rust
let flow_inits = MaybeInitializedPlaces::new(tcx, &body, &mdpe);
let mut flow_inits = dataflow::generic::Engine::new_gen_kill(tcx, &body, def_id, flow_inits)
        .iterate_to_fixpoint()
        .into_cursor(&body);
```

could instead be written like this:

```rust
let mut flow_inits = MaybeInitializedPlaces::new(tcx, &body, &mdpe)
        .into_engine(tcx, &body, def_id)
        .iterate_to_fixpoint()
        .into_cursor(&body);
```

Another idea was to add `HasBody` as a supertrait of `AnalysisDomain`. This would obviate the need to pass `body` everywhere. With the current API, it's possible (albeit unlikely) for a user to pass a different `mir::Body` into, for example, `into_engine` than was used to create the analysis, which would cause subtle bugs. I believe this would only work if the `Analysis` methods used `&self`, however.

ecstatic-morse commented:
@pnkfelix I addressed the motivation for the `GenKill` trait and added some documentation/renamed some methods in the last two commits.

pnkfelix commented:
Thanks @ecstatic-morse !

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 7b4dca282a0bfad265238d04b22f7bdb0d498d74 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 7b4dca282a0bfad265238d04b22f7bdb0d498d74 -->
<!-- homu: {"type":"Approved","sha":"7b4dca282a0bfad265238d04b22f7bdb0d498d74","approver":"pnkfelix"} -->

bors commented:
:hourglass: Testing commit 7b4dca282a0bfad265238d04b22f7bdb0d498d74 with merge 2cf24ab89123532f0e84995301ee734a88bf4701...
<!-- homu: {"type":"BuildStarted","head_sha":"7b4dca282a0bfad265238d04b22f7bdb0d498d74","merge_sha":"2cf24ab89123532f0e84995301ee734a88bf4701"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=18987)
Approved by: pnkfelix
Pushing 2cf24ab89123532f0e84995301ee734a88bf4701 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=18987"},"merge_sha":"2cf24ab89123532f0e84995301ee734a88bf4701"} -->
